### PR TITLE
Refactor low-level layer operations

### PR DESCRIPTION
**Pull request by Malax**
This PR refactors the internal functions in `layer/handling.rs` in preparation of a new Layer API PR. The only public (and strictly speaking breaking) change are the new error enums for writing SBOM and exec.d. 

The initial PR that also exposed these low-level building blocked proved to be more controversial than I expected so this PR changed to only refactor the internal bits.

## Old PR description

~~This PR is the first phase for a new high-level Layer API. It exposes the low-level building blocks needed to work with CNB layers. It is **not** the new layer API.~~

~~Why do we need this when we're aiming for a new _high-level_ API?~~

~~First of all, that new high-level API needs access to these operations. They were originally written as helpers for the `Layer` trait API and and needed some adjustments to make them usable in other contexts. I mainly improved error handing and naming. This work needed to be done in any case and this PR front loads these changes in a separate PR for easier review. :)~~

~~Secondly, regardless of how great the new high-level API will be, there will always be fringe use-cases where low-level building blocks are needed. Right now, users of libcnb.rs are out of luck if they ever hit these uses-cases. After this PR, we have an escape hatch when a higher-level abstraction (either the current or new one) doesn't work.~~

## ~~Notes for reviewers~~
~~Start with the `BuildContext` changes in [libcnb/src/build.rs](https://github.com/heroku/libcnb.rs/pull/786/files#diff-44c804e2de395175db23550b14a017b1cca0d24c8b3cd4b314c7afdca643dacf). This is the newly exposed API for users. Everything else is just supporting changes.~~

~~I've broken up the steps I took in separate commits. It might be helpful to refer to those to see how things changed in atomic steps from the original.~~

### CODE REVIEW

File: libcnb/src/layer/handling.rs
Comment (schneems) [inline libcnb/src/layer/handling.rs:387]: Variable names `layers_dir` and `layer_dir` are very close to my eyes. Here you're introducing a variable to use it only in the next function call below. I think it would be less confusing to inline all that logic and remove the variable.


```rust
    layer_env.write_to_layer_dir(layers_dir.join(layer_name.as_str()))?;
```

Alternatively, `write_to_layer_dir` would be safer if it took a path to the layers dir and a `LayerName` and did this concatenation for you.

Comment (Malax) [reply libcnb/src/layer/handling.rs:402]: We have over 100 occurrences of `layer_dir` and `layers_dir` variables in this project. We actually disable a specific lint in `handling.rs` to allow for this: 

```
// This lint triggers when both layer_dir and layers_dir are present which are quite common.
#![allow(clippy::similar_names)]
```

Regardless of how severe we think this this issue is, it's a not a problem of this PR I would say. If we feel strongly about this naming we should address it holistically. Personally, I don't think having `layers_dir` and `layer_dir` is a problem, but I can see your argument of them being similar.

Comment (schneems) [reply libcnb/src/layer/handling.rs:402]: > Regardless of how severe we think this this issue is, it's a not a problem of this PR I would say. If we feel strongly about this naming we should address it holistically. Personally, I don't think having layers_dir and layer_dir is a problem, but I can see your argument of them being similar.

When dealing with paths it's easy to confuse passing in a directory you need versus passing in the parent directory of the directory you need. The naming is a piece of the puzzle. My larger worry would be that the types don't protect us in this situation.

If you're fine with it as-is, I'm fine with it. I'm poking things and seeing how they move.

**Review by schneems — CHANGES_REQUESTED** [review]
## Requesting:

- There's no way to write an environment variable that's exposed right now. I would like one to be added.
- Sboms and execd can be written but not read (so not updated). I would like a way to read these values.

## Proposing

I left some comments about possible invalid state if launch metadata isn't written. I know we've said that there will be a higher level interface for v2 but I'm warry about needing a function to expose each of the pieces of functionality. These associated functions, while not recommended, are on the BuildContext so users will certainly find them.

I think it's possible and preferable to reduce all these functions to one write and one read function. An update can be composed of a read followed by a modification and a read. Ideally the type that is returned by the read can be easily converted (with minimal extra data such as `LayerTypes`) into the type needed to write.

Such an "atomic" set of read/write operations would remove the need for checks and error types like https://github.com/heroku/libcnb.rs/pull/786/commits/2a6bc5502822b9181f21fbeb991c457387bd740a. 

An alternative could be requiring `LayerTypes` input on every function call. Though I like that less, as there's an open question of what should happen in case the programmer issues two writes with different values.

File: libcnb/src/build.rs
Comment (Malax) [reply libcnb/src/build.rs:118]: This PR is really just about exposing what we currently have. There is no built-in way to retrieve exec.d and SBOM information currently (beyond the types in `libcnb-data` and testing utils). I'd like to keep this PR scoped to that and do follow-up work separately. Maybe opening an issue for adding these facilities (which would also be valid without this PR as there is no way to do this right now either).

File: CHANGELOG.md
Comment (Malax) [reply CHANGELOG.md:17]: That is correct. See the PR description: 

> Why do we need this when we're aiming for a new high-level API?
>
> First of all, that new high-level API needs access to these operations. They were originally written as helpers for the Layer trait API and and needed some adjustments to make them usable in other contexts. I mainly improved error handing and naming. This work needed to be done in any case and this PR front loads these changes in a separate PR for easier review. :)
>
> Secondly, regardless of how great the new high-level API will be, there will always be fringe use-cases where low-level building blocks are needed. Right now, users of libcnb.rs are out of luck if they ever hit these uses-cases. After this PR, we have an escape hatch when a higher-level abstraction (either the current or new one) doesn't work.

File: libcnb/src/build.rs
Comment (Malax) [reply libcnb/src/build.rs:135]: > Usually the buildpack maintainer creates these things independently

They are stored in the same file ([CNB spec](https://github.com/buildpacks/spec/blob/7eb38d15fd3bd5bb5240584f0e70beb61913032c/buildpack.md#layer-content-metadata-toml)) that's why we have `LayerContentMetadata` in `libcnb-data`. How would a buildpack maintainer create these independently? Are you referring to the existing `Layer` trait API? Things like this are the reason why we have a higher-level API to make working with layers more palatable. But the low-level API dictated by the spec is what it is.

> This also makes me realize that the current state of this PR has layers that might have execd or sboms but have no launch/build/cache metadata by accident.

That's how it work with CNBs - it's just stuff on a file system. The low-level API here is not trying to hide any of that by design. The new SBOM and exec.d methods on `BuildContext` at least check the layer exists before writing either, but that's the extend of what we can do.

File: libcnb/src/build.rs
Comment (Malax) [reply libcnb/src/build.rs:117]: I'm not 💯 sure what you mean. My guess is that you want to remove the `Option` and make a non-existent layer an error variant of `ReadLayerError`? Because it is already is a `Result`.

In the end it's really just the decision if we consider a missing layer an error or a valid and expected outcome of this function. You can find arguments for both and I think there is no clear winner. Which solution is the "better" solution heavily depends on context.

For example, for `std::fs`, a missing file is an error which I think is reasonable as this mirrors what the operating systems do when accessing a missing file. The other extreme would be a collection API, where you wouldn't expect a `Result` for retrieving a value by key from a hash map.

In this case, I chose to signal a non-existing layer via an `Option`. However, for this function to operate, I have to parse data and do IO. As these operations are fallible, I need to bubble them up and wrap the result of the function as a `Result`. This make the final return type a `Result<Option<_>, _>`.

> Introducing a third would mean one fewer operation to unwrap and get the contents.

I think the ergonomic of that return type are alright. Some examples:

When you really care about all cases, layer existing, layer missing, specific errors and a catch-all for all other errors:

```rust
match result {
    Ok(None) => /* */,
    Ok(Some(layer)) => /* */,
    Err(SpecificError) => /* */,
    Err(_) => /* */
}
```
When you only care about the layer being there or not:

```rust
if let Some(layer) = read()? {
    /* */
} else {
    /* */
}
```

When you absolutely need the layer to exist and you have no way to recover (I don't think this will ever be the case for a real-world buildpack):

```rust
let layer = read()?.ok_or(BuildpackError::IReallyNeedThisLayer)?;
```

I'm not convinced that this needs changing, but happy to be convinced otherwise either by you or other folks :)

File: libcnb/src/layer/handling.rs
Comment (Malax) [reply libcnb/src/layer/handling.rs]: Can you give a reasoning why you would prefer `impl` here?

Comment (schneems) [reply libcnb/src/layer/handling.rs]: Mentally the two are synonymous, but `impl TraitName` is easier for me to read. 

Here's roughly how I translate the experiences of reading those two different styles. 

- `layers_dir: impl AsRef<Path>: "I saw a dog that is blue" has all of the information I need in the order I need 
- `<P: AsRef<path>>`  ... `layers_dir: P` "I saw a dog B" ... look for the definition of B going backwards "B is blue" then put them back together "I saw a dog B, which means it is blue."  Or alternatively "B will mean blue" okay, I'll hold onto this and don't know when it will be useful..."I saw a dog B" oh right, I remember that I know what B is already "therefore it is blue." 

It's the difference between reading `3 5 +` (reverse polish notation) and `3 + 5`. Both have the same information, but one requires more temporary memory to grok.

I was curious if there are practical implications towards using this versus a generic so I looked up the differences and found this https://stackoverflow.com/a/47515540. The downside is that you cannot explicitly specify the type in the function signature, which I think is perfectly fine here.

File: libcnb/src/build.rs
Comment (schneems) [reply libcnb/src/build.rs:117]: > My guess is that you want to remove the Option and make a non-existent layer an error variant of ReadLayerError? Because it is already is a Result.

Yes, exactly. I was considering that out loud. I came to the conclusion that I think I actually like the `Result<Option>` 

> I'm not convinced that this needs changing, but happy to be convinced otherwise either by you or other folks :)

Good. I'm of the same persuasion. I left the comment to have it validated. We can resolve this.

File: libcnb/src/layer/handling.rs
Comment (edmorley) [reply libcnb/src/layer/handling.rs]: I find the `impl` version easier to read too fwiw.

I kinda see the `P: AsRef<Path>` notation as an advanced notation needed only in certain situations (where `P` is needed in multiple places etc), and so in general when writing things like this I would lean towards using the `impl` form unless it otherwise wasn't viable.

File: CHANGELOG.md
Comment (schneems) [reply CHANGELOG.md:17]: Got it. I'm onboard, though with some caveats.

Short: I think we should error when `LayerType`-s are not configured for functions: `replace_layer_exec_d_programs`, `replace_layer_sboms` and `EnvLayer::write_to_layer_dir`

Long:
----
Rather than a minimum API, I'm interested in a "minimum safe API". My major fear is that something is written without a `LayerType` in which case it would look like it's having some effect, but is silently having no effect. Which, if we make it ergonomic and possible to do, it will happen.

## What I like in this PR:

- `read_layer` any read operation won't have this problem 👍🏻
- `write_layer_metadata` by definition won't have this problem 👍🏻
- `delete_layer` listed for completeness 👍🏻

## What I don't like

- `replace_layer_exec_d_programs` 
- `replace_layer_sboms`
- `EnvLayer::write_to_layer_dir`

These all allow writing values to layers that may or may not do what the programmer expects.

## Options

- Explicitly checking that LayerType data is written to the layer on each of these API calls and returning an error if it's not. If a programmer doesn't care, they can catch and disregard the error.
- A "kitchen sink" read/write minimal API. One that accepts every possible input to a layer to force the programmer to consider the LayerType. This is more-or-less what I was advocating for before, what you're calling a "higher level" API.
- Requiring each of these functions to accept a LayerType

## Evaluation

I think the most "rust" way to handle this concern would be checking and erroring if the LayerType data is not set. It won't fail a compilation when the programmer makes a mistake, but at least it will be caught in an integration test instead of failing silently and in unexpected ways. It's a little more than "minimal" API, but I think it's within "minimal safe API."

## Suggestion

Check correct LayerType information is present and error if missing for:

- `replace_layer_exec_d_programs` 
- `replace_layer_sboms`
- `EnvLayer::write_to_layer_dir`

While `EnvLayer::write_to_layer_dir` is not in your PR, it already returns a Result and I think we should take the opportunity to standardize.

I would go a step further for ergonomics/consistency and say we should have an api `fn replace_layer_envs` on context that takes an `EnvLayer` so programmers don't miss this functionality (like I did).

File: libcnb/src/build.rs
Comment (schneems) [reply libcnb/src/build.rs:118]: I'm fine with that as long as we know we're not being exhaustive now.

File: libcnb/src/build.rs
Comment (schneems) [reply libcnb/src/build.rs:135]: Short: I withdraw my suggestion here, you can close this comment. 

Long:
----
I was about to ask for a doc test to show users how to construct that type, but realized it's a pain in the butt to doctest things inside of the build context. So for my own understanding here's what the function call would look like:

```rust
write_layer_metadata(
    layer_name!("trying_this_on"),
    LayerContentMetadata{
        types: Some(LayerTypes {
            build: true,
            launch: true,
            cache: true
        }),
        metadata: my_metadata,
    }
);
```

Which is not that different than accepting two arguments.

**Review by schneems — CHANGES_REQUESTED** [review]
I've commented on each of your comments. Two of them can be closed. For the third:

I think we should error when LayerType-s are not configured for functions: `replace_layer_exec_d_programs`, `replace_layer_sboms` and `EnvLayer::write_to_layer_dir`

File: libcnb/src/layer/handling.rs
Comment (Malax) [reply libcnb/src/layer/handling.rs:376]: Since we're talking preferences here and this PR is not changing how we do these things in this existing file - can we move this discussion/change request of of this PR?

**Comment by Malax** [comment]
@schneems, @edmorley since the exposure of the low-level building blocks seems much more controversial than I expected, those changes have been removed from this PR. It's now only refactoring the internal functions and slightly changes `WriteLayerError`.

I still think we want to do this eventually (and some of it is needed for the new Layer API), but I want the internal changes merged ASAP to be able to open the actual new Layer API PR.

I will re-request review from you both. :)

File: libcnb/src/layer/handling.rs
Comment (edmorley) [inline libcnb/src/layer/handling.rs:246]: (optional)
Should `WriteLayerError` be listed above the other enums here, given it's the topmost enum in the hierarchy?

**Review by edmorley — APPROVED** [review]

File: libcnb/src/layer/handling.rs
Comment (schneems) [reply libcnb/src/layer/handling.rs:246]: I don't have a strong opinion about the location. For consistency, putting it with the other sub-errors seems good though.

**Review by schneems — APPROVED** [review]
I see two changes: 

- Rename "overwrite" to "replace"
- Extract functionality from `write_layer` to their own methods.

And an associated change to the error enum as documented in the CHANGELOG.

None of these changes are controversial. I'm happy to move forward. I'm okay moving the location of the enum. I would like docs for completeness sake and left a suggestion, but I also won't block on that either.

File: CHANGELOG.md
Comment (schneems) [reply CHANGELOG.md]: Since these have been ripped out. I'll dismiss this conversation.

File: libcnb/src/layer/handling.rs
Comment (Malax) [reply libcnb/src/layer/handling.rs:237]: Also no strong opinion here, I reordered them in c12b1b0fdc2a82368b1a97fe0020b3e75a4f5ec0

File: libcnb/src/layer/handling.rs
Comment (Malax) [reply libcnb/src/layer/handling.rs]: This seems out of scope for this PR - do you want to open an overall libcnb issue for it?

**Pull request merged**
